scripts/circleci_usage_data.py

```diff
@@ -0,0 +1,267 @@
+from scalyr_agent import compat
+
+if False:
+    from typing import Optional
+    from typing import List
+
+import sys
+import argparse
+import datetime
+
+from io import StringIO
+
+import requests
+
+# See https://circleci.com/pricing/#comparison-table
+CREDIT_BUNDLE_PRICE = 15.0
+CREDIT_BUNDLE_UNIT_COUNT = 25000.0
+
+# Price for a single CPU usage credit
+PRICE_PER_CREDIT = CREDIT_BUNDLE_PRICE / CREDIT_BUNDLE_UNIT_COUNT
+
+GITHUB_API_PRS_URL = "https://api.github.com/repos/{project_slug}/pulls"
+GITHUB_API_SEARCH_URL = "https://api.github.com/search/issues"
+CIRCLE_CI_API_INSIGHTS_URL = (
+    "https://circleci.com/api/v2/insights/{project_slug}/workflows/{workflow}"
+)
+CIRCLE_CI_WORKFLOW_VIEW_URL = "https://circleci.com/workflow-run/{workflow_id}"
+
+CIRCLE_CI_API_TOKEN = compat.os_environ_unicode.get("CIRCLE_CI_API_TOKEN", None)
+MAILGUN_API_TOKEN = compat.os_environ_unicode.get("MAILGUN_API_TOKEN", None)
+
+
+def get_all_branches_for_repo(project_slug):
+    # type: (str) -> List[str]
+    """
+    Retrieve all the branches (including deleted ones) for a particular repo.
+
+    NOTE: This function utilizes pulls API endpoint since /branches one doesn't include deleted
+    branches.
+    """
+    # 1. Retrieve all the PRs for last week
+    url = GITHUB_API_SEARCH_URL.format(project_slug=project_slug)
+
+    now = datetime.datetime.now().strftime("%Y-%m-%d")
+    week_ago = (datetime.datetime.now() - datetime.timedelta(days=7)).strftime(
+        "%Y-%m-%d"
+    )
+
+    query = "repo:{project_slug} is:pr created:{week_ago}..{now}".format(
+        project_slug=project_slug, week_ago=week_ago, now=now
+    )
+    params = {"q": query, "per_page": "1000", "sort": "created"}
+
+    response = requests.get(url, params=params)
+    data = response.json()
+    pull_numbers = [item["number"] for item in data["items"]]
+
+    # 2. Retrieve all the branches and filtered ones which haven't been updated during last week
+    # (based on the PR activity)
+    # 1. Retrieve all the branches
+    url = GITHUB_API_PRS_URL.format(project_slug=project_slug)
+    params = {"per_page": "100", "sort": "created", "direction": "desc", "state": "all"}
+    response = requests.get(url, params=params)
+    data = response.json()
+
+    branches = []
+    for item in data:
+        branch_ref = item.get("head", {}).get("ref", None)
+        pull_number = item["number"]
+
+        if not branch_ref or pull_number not in pull_numbers:
+            continue
+
+        branches.append(branch_ref)
+
+    return branches
+
+
+def get_usage_data_for_branch(
+    buff, project_slug, workflow, status="success", branch="master", limit=10
+):
+    # type: (StringIO, str, str, str, str, int) -> None
+    """
+    Get usage data for the provided project and write it to the provided buffer.
+    """
+    url = CIRCLE_CI_API_INSIGHTS_URL.format(
+        project_slug=project_slug, workflow=workflow
+    )
+    params = {"limit": "100", "branch": branch}
```

Comment:
Why not use the value from the `limit` argument here?

Reply:
The limit as in the command line option refers to the maximum number of workflow run we want to see for each branch and the API endpoint returns runs for all the workflows.

Right now that API endpoint doesn't support per-workflow filtering so we just request some large number of items and do that late filtering in the Python code (and we also do some additional late filtering based on the workflow status, etc.).

So yeah, that approach is not 100% ideal and robust since we may actually miss some data (depending on the filters and actual data returned by the API), but that's good enough for this use case, imo (and that API endpoint doesn't support pagination either).